### PR TITLE
Enhance instrumentation docs in Phoenix.Instrument

### PR DESCRIPTION
**Pull request by whatyouhide**
This PR adds documentation to the `Phoenix.Endpoint.Instrument` module which gives a detailed overview of the instrumentation API we added to Phoenix. Let me know how it looks!

### CODE REVIEW

**Comment by josevalim** [comment]
This is a private module, we should not expose it. We can either add this
to Phoenix.Endpoint (which I think we did?) or move it to a guide?

Other than that, the docs are excellent! :D


-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Director of R&D

**Comment by chrismccord** [comment]
Ah sorry, I told @whatyouhide to open up the module with a `@moduledoc`. We can put on in `Phoenix.Endpoint`, but it's a quite a chunk of docs. I'd prefer to see it live in the codebase vs a separate guide. This would be easier once we have everything on ex_doc, but if it's b/w guide or docs, let's move the docs to `Phoenix.Endpoint` with  an Instrumentation section

**Comment by whatyouhide** [comment]
I completely overlooked the fact that I already wrote quite a lot of docs in `Phoenix.Endpoint`, so lots of things were duplicated between `Phoenix.Endpoint` and `Phoenix.Endpoint.Instrument`. I moved all the docs to `Phoenix.Instrument`, so this should be ready for review.

File: lib/phoenix/endpoint.ex [hunk 1]
```diff
@@ -240,30 +240,81 @@ defmodule Phoenix.Endpoint do
...
+            end
+
+            def event_callback(:stop, _time_diff, start_time) do
+              stop_time = :erlang.monotonic_time(:micro_seconds)
+              do_something_with_diff(stop_time - start_time)
+            end
+          end
+
+  The return value of each "before" event callback will be stored and passed to
+  the corresponding "after" callback.
```
Comment (josevalim) [inline lib/phoenix/endpoint.ex]: Should we add a tiny note saying it is not guaranteed the after event will be invoked? And also say the instrumenter is invoked in the same process as the one running the request so we should be careful with blocking actions?

**Comment by josevalim** [comment]
One tiny suggestions and we are good to go!

File: lib/phoenix/endpoint.ex (hunk 1 above)
Comment (whatyouhide) [inline lib/phoenix/endpoint.ex]: +1 on the same process thing, but I think we always call "after" callbacks? https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/endpoint/instrument.ex#L67-L72

Comment (josevalim) [inline lib/phoenix/endpoint.ex]: We don't always call because a link may break.

**Comment by whatyouhide** [comment]
@josevalim made the changes, ready for review

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by chrismccord** [comment]
Thanks Andrea!

**Pull request merged**